File: solutions/node/src/charging/application/ports.ts
import type { ChargingSession } from '../domain/ChargingSession.ts';

// Ports (hexagonal): the application layer depends on these interfaces,
// adapters (in-memory today, SQLite/Postgres on Day 6) implement them.
export interface ChargingSessionRepository {
  findById(sessionId: string): Promise<ChargingSession | undefined>;
  findActiveByConnector(connectorId: string): Promise<ChargingSession | undefined>; // status === Charging
  save(session: ChargingSession): Promise<void>;
}

export interface AuthorizationService {
  isAuthorized(idTag: string): Promise<boolean>;
}

File: solutions/python/src/charging/application/ports.py
"""Ports (hexagonal): the application layer depends on these Protocols,
adapters (in-memory today, SQLite/Postgres on Day 6) implement them."""
from typing import Protocol

from charging.domain.charging_session import ChargingSession


class ChargingSessionRepository(Protocol):
    def find_by_id(self, session_id: str) -> ChargingSession | None: ...
    def find_active_by_connector(self, connector_id: str) -> ChargingSession | None: ...  # status == Charging
    def save(self, session: ChargingSession) -> None: ...


class AuthorizationService(Protocol):
    def is_authorized(self, id_tag: str) -> bool: ...
